### ISSUE TITLE
Flow Server does not work with JBoss EAP 7.0 after 14.1.24

### ISSUE THREAD

**Issue by werwolfberlin** [vaadin/flow#8173]
There is an issue running the Vaadin application on a JBoss 7.0 EAP. Following observations have been made:

initial run after "mvn clean" and wiping the node_modules folder seems to be successful. "mvn install" and deploy worked, but **the UI of the application is incomplete** (only header). All following "mvn install" with a followed deploy ended with an error in the "install" part:

```
[ERROR] Failed to execute goal org.wildfly.plugins:wildfly-maven-plugin:1.2.1.Final:deploy (default) on project xyz-vaadin-delivery: Failed to execute goal deploy: {"WFLYCTL0062: 
Zusammengesetzte Operation ist fehlgeschlagen und wurde zurückgesetzt. Die Schritte die fehlgeschlagen sind: " => {"Operation step-1" => {"WFLYCTL0080: Fehlgeschlagene Dienste" => {"jboss.undertow.deployment.default-server.default-host./xyz-vaadin-mock" => "org.jboss.msc.service.StartException in service jboss.undertow.deployment.default-server.default-host./xyz-vaadin-mock: java.lang.RuntimeException: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
[ERROR]     Caused by: java.lang.RuntimeException: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
[ERROR]     Caused by: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
[ERROR]     Caused by: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib"}}}}
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.wildfly.plugins:wildfly-maven-plugin:1.2.1.Final:deploy (default) on project xyz-vaadin-delivery: Failed to execute goal deploy: {"WFLYCTL0062: Zusammengesetzte Operation ist fehlgeschlagen und wurde zurückgesetzt. Die Schritte die fehlgeschlagen sind: " => {"Operation step-1" => {"WFLYCTL0080: Fehlgeschlagene Dienste" => {"jboss.undertow.deployment.default-server.default-host./xyz-vaadin-mock" => "org.jboss.msc.service.StartException in service jboss.undertow.deployment.default-server.default-host./xyz-vaadin-mock: java.lang.RuntimeException: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
    Caused by: java.lang.RuntimeException: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
    Caused by: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
    Caused by: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib"}}}}
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to execute goal deploy: {"WFLYCTL0062: Zusammengesetzte Operation ist fehlgeschlagen und wurde zurückgesetzt. Die Schritte die fehlgeschlagen sind: " => {"Operation step-1" => {"WFLYCTL0080: Fehlgeschlagene Dienste" => {"jboss.undertow.deployment.default-server.default-host./xyz-vaadin-mock" => "org.jboss.msc.service.StartException in service jboss.undertow.deployment.default-server.default-host./xyz-vaadin-mock: java.lang.RuntimeException: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
    Caused by: java.lang.RuntimeException: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
    Caused by: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
    Caused by: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib"}}}}
    at org.wildfly.plugin.deployment.AbstractDeployment.execute (AbstractDeployment.java:152)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :xyz-vaadin-delivery
```

The essence of this is:
`java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib`

As you can see: it is a Windows machine.

The Code runs in a JBoss EAP 7.0 standalone which is a separate process. Deploy triggers the server. this might lead to the non empty directory, because one server process could have it opened, so it could not be deleted.

- Versions:
    - Vaadin version: 14.1.24/25/26
    - Java version: 1.8.0_241
    - OS version: Windows 10 Pro
    - Browser version: Chrome, but does not matter

**Comment by denis-anisimov** [comment]
As I understand you are using dev mode with deploying an app to the external server : JBoss.

Do you really want to use dev mode ? Is this done intentionally ?

Windows has a known issue : it may not remove files if they are used by another process.
So : _we may not do anything with this_. We may not remove files on Windows if they are locked by another process. It's just impossible.

So I see only two ways here:
- use production mode which doesn't lock any files since it doesn't need read anything from FS
- stop and start the server every time when you want to use `mvn deploy` .

In fact if you are using maven for deploying I don't see a point to use dev mode.

Otherwise it has sense to use "Run on Server" in your IDE instead maven.

As you can see it's not even Vaadin which tries to remove the file : it's maven
```
\polymer\\lib
    Caused by: java.io.UncheckedIOException: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib
    Caused by: java.nio.file.DirectoryNotEmptyException: D:\\Projects\\xyz-vaadin\\xyz-vaadin-samples\\xyz-vaadin-delivery\\node_modules\\@polymer\\polymer\\lib"}}}}
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
```
But yes, files are locked because of Vaadin code  in dev mode running on server.

**Comment by werwolfberlin** [comment]
Hi Denis,

I am aware that there is a workaround.
I want to reflect a good picture of vaadin to the customer who is making a
decision about a contract of many many hours, but I cannot achieve that
when I/they have trouble like that. Currently we are running 14.1.20 with a
pinned flow server 2.2 snapshot (which at least applies js files but no css
files) and a pinned combobox version which has been fixed for Prevo. When
we want to deploy css changes, we run TommEE, because it is repairing
things which the JBoss messes up. I am not able to hot-deploy JS or CSS
with JBoss, except when I run TomEE WHILE I am running JBoss (which is also
just another build). But that procedure often ends up in ghost processes
which then lock again more files.

When i switch to 14.1.25 i need to clean everything up and even though
the first run is free of errors, it does only show 4 buttons instead of the
whole site. a re-run ends with that file access error which i don't have
with flow-server snapshot 2.2.

An install run takes about 2 minutes on a 4GHz Xeon with 6C/12T. Currently
we have about 15 views developed for the POC. There will be more than 500
views done with the designer (JS + CSS). I can just guess how many
times it will take to build and deploy the app with 500 views.

A lot of machines in the IT industry are windows. I personally didn't have
any customers yet which used Mac or Linux.

so my goal is:
- make hot-deploy work to reduce development cycles
- represent a good picture of vaadin

I hope I could make myself understood. It would be super nice if i get some
help here. I have to work with the designer team as well to make their
product a bit more ripe and struggle also with other issues. Coding and
running after things which don't work does not leave me much time for the
real work with the customer. I am also open for help how to configure that
JBoss server in IntelliJ so that I am able to debug. At the moment i am
attaching a debug process to the debug port of JBoss.

Please help me to make vaadin shine a bit in front of the customer.

Happy Weekend!

Gerald


Am Do., 30. Apr. 2020 um 15:27 Uhr schrieb Denis <notifications@github.com>:

**Comment by denis-anisimov** [comment]
>I am aware that there is a workaround.

Sorry, which workaround you mean ?

>A lot of machines in the IT industry are windows.

Sure, no doubts that Win is quite popular platform for users.

I still don't understand the required scenario.

Let's fix the facts:
- Win doesn't allow to remove the files which are used by another process.
- If you are running dev mode then frontend files are locked by web-pack process (this is not even Vaadin this is _web-pack_ `npm` process which is executed inside the web server in dev mode).
- `maven deploy` goal tries to remove some files which are locked ( I'm not sure why be the way: do you understand why it tries to remove them ?) 

So I don't see what we may do here: this is how dev mode works.

Normally when you deploy to external web server dev mode should not be used.

You are using maven for deploy anyway. So is there any specific reason to use dev mode + `maven deploy` instead build production web app and deploy it ?

Dev mode allows automatically rebuild frontend resources if they are changed without redeploying. But you deploy anyway: so I don't see a reason do use dev mode then.
Am I missing something from the scenario ?

If the main goal to speed up development then:
- use dev mode (for frontend resources)
- don't use maven for deploy
- use either IDE functionality (like "Run on Server" in Eclipse). IDE integration in many cases allows to do hot deploy of the binaries without need to redeploy
- or use some live reload tools . You can find documentation about that here: https://github.com/vaadin/flow-and-components-documentation/blob/master/documentation/workflow/workflow-overview.asciidoc

Unfortunately I don't use InelliJ IDEA, so I can't help you here with it configuration.
But regardless of IDE : hotswap agent or JRebel allows to reload Java code without redeployment.

The avoidance of redeployment is the main purpose of Dev mode.
So as I understood your scenario it's just an attempt to use Dev mode in an "incorrect" way, the way which it's not supposed to be used: either deploy (and then use production mode) or avoid deployment (and then use dev mode).


Have a nice holidays too!

**Comment by werwolfberlin** [comment]
> Sorry, which workaround you mean ?
The production mode

i will try your linked document and see if i can find some things which are applicable to our setup.

The thing which i don't understand is: why is snapshot 2.2 running without that locking issue and why can i run TomEE WHILE i have deployed stuff in JBoss and there is no lock issue and it is updating the CSS also in the JBoss deployment (14.1.20 without the 2.2 snapshot)?

**Comment by denis-anisimov** [comment]
Good question.
I also don't know.

I don't remember all the decisions but may be `2.2` uses `pnpm` for packaging ?
Are you able to check whether `pnpm` or `npm` is used with `2.2-SNAPSHOT`?
You may try to specify explicitly `npm` for `2.2-SNAPSHOT` and see what happens.

Another question is : why deployment tries to remove the folder.
I don't understand a single reason for this behavior.
It seems that the project has some specific configuration which considers `node_modules` folder as required to be removed on deployment.
As I understand : there are no reasons to remove this folder ever (unless you specifically want to do clean deploy).
The exception comes from wildfly deploy plugin:
`org.wildfly.plugin.deployment.AbstractDeployment.execute (AbstractDeployment.java:152)`.
I suspect that deploy in wildfly plugin doesn't remove all folders in the project by itself ( and of course `node_modules`  can't be hardcoded in the plugin): so there is something in `pom.xml` which says that this folder needs to be removed for some reason.
And this configuration is the reason of the behaior.

**Comment by werwolfberlin** [comment]
OK, here some sumarized conclusions:
the problem is: Running the app (CDI) on JBoss with Vaadin 14.1.24+ under windows in dev mode does not work anymore due to
a) incomplete deployment
b) file lock exception on redeploy

**Comment by denis-anisimov** [comment]
So just to make sure we are on the same page: I'm waiting the project which I can use to reproduce the issue.

**Comment by werwolfberlin** [comment]
yes, i delegated it to a colleague since i have to finish stuff for the
customer.

Am Mo., 11. Mai 2020 um 09:33 Uhr schrieb Denis <notifications@github.com>:

**Comment by denis-anisimov** [comment]
Alright 
👍

**Comment by enver-haase** [comment]
That would be me, sorry about the delay. At it.

**Comment by enver-haase** [comment]
It appears to me as if the two of you meant different things when talking about deployment (mvn deploy:deploy vs mvn wildfly:deploy) or "hot" deployment (redeploying an artefact while a server is running vs hot-swapping/hot-patching code in a running server without redeploying).

This comment is mostly here to say 'I cannot reproduce the problem yet, discussing with OP @werwolfberlin '. Please stay tuned.

**Comment by enver-haase** [comment]
I still cannot reproduce this. Also I experimented with the Maven Wildfly plugin and I can easily deploy a development version using mvn wildfly:run and mvn wildfly:start / mvn wildfly:deploy / mvn wildfly:shutdown -- in each case changing CSS in the source code directory was picked up immediately in the Chrome browser when reloading the page (cache being turned off).

I am closing this as "cannot reproduce".

This bug was meant to help making sure we have low turnaround times for upcoming tasks involving updating CSS and JS code on-the-fly and I am certain now that we will be able to achieve this.
Tested only with wildfly-maven-plugin:2.0.2.Final (using wildfly-19.1.0.Final as its default) though.
